---
title: "BMS 270/BMI 219 The Human Microbiome (2018)"
date: "May 15, 2018"
output: 
  html_document:
    fig_width: 11
    fig_height: 8.5
---

```{r, include=F}
knitr::opts_chunk$set(echo = TRUE, message=TRUE, warning=FALSE,  cache=F, tidy=TRUE)
```
# Computational Methods for Exploring the Microbiome
***

The following tutorial focuses on a subset of data derived from: [The microbiota at multiple body sites during pregnancy in a rural Tanzanian population and the effects of Moringa supplemented probiotic yogurt doi:10.1128/AEM.00780-15](http://aem.asm.org/content/early/2015/05/13/AEM.00780-15.abstract). The study was comprised of a longitudinal analysis of 56 pregnant women. We will revisit the finding that there is a major shift in the vaginal microbiome at birth using a completely independent set of tools from those in the paper using a workflow entirely carried out in R. We will focus on a subset of 20 women for the purposes of this analysis comparing enrollment in the second trimester, to birth. We are using the vaginal microbiome data set here as it has relatively lower diversity compared to the gut and in previous analysis showed very clear shifts in composition.


The tutorial will require some basic knowledge of the R language. At this point in time, knowing bash, R, python, matlab, and/or some combination thereof is required in the microbiome field. All languages use more or less the same basic logic (for biological computing anyway) but differ in their syntax. It is more important to know what you want to do, than how to do it. Do not hesitate to ask for help if you are having trouble. Also check [here](https://www.rstudio.com/resources/cheatsheets/) for helpful cheat sheets. The commands you will need are all supplied in gray boxes with the output below.
<br>

If you are not familiar with R, please watch these videos:
*  [Introduction to R Programming with DataCamp](https://www.youtube.com/watch?v=HkNFn6eosaU)
*  [Getting started with R and RStudio](https://www.youtube.com/watch?v=lVKMsaWju8w&t=3s)
*  If you are interested in learning more R, I recommend [this book](https://www.amazon.com/Learning-Step-Step-Function-Analysis/dp/1449357105).

***

# Installing necessary requirements {#Installation}

***

## This entire process could take ~10-15 minutes, please ensure this has been done before coming to class!

If you do not already have R, download the correct version for your operating system [here](https://cran.cnr.berkeley.edu/). Then download the newest version of R studio for your operating system [here](https://www.rstudio.com/products/rstudio/download2/). Please ensure your version of R is at least 3.4.2 . You can check this with the following command:

```{r}
print(R.Version()$version.string)
```

<b>If you are a mac user who has never used any form of bioinformatic tool or programming language on your computer, it may be a good idea to download the Xcode Developer tools from the Apple app store</b>

Open R studio and click File > New File > R script. Save this file in a folder of your choice as <i>YourName.MicrobiomeTutorial.R</i>.

Next, click Session > Set Working Directory > Choose Directory. Select the folder in which you placed your .R file.

### For Ubuntu users only:

```{r, eval=F}
###For Ubuntu users only, you may need to do the following:
sudo apt-get install libcurl4-gnutls-dev #in terminal
sudo apt-get install curl #in terminal
sudo apt-get install gsl-bin libgs10-dev #in terminal
install.packages("RCurl") #in R
```

Next run the following code which will download all required packages for this tutorial if you do not already have them. It will also download the raw data from European Nucleotide Archive. <b>If asked to compile, reply yes</b>, this may happen multiple times. On OSX you may be asked to install developer tools to which you should accept. <b>If asked to update other packages, type n for none</b>. You may have to do this multiple times. <i>For advanced users, this script may update versions of packages you already have. Feel free to manually install or update as needed based on the source code of the install script.</i>

```{r, eval=F}
source("https://raw.githubusercontent.com/jbisanz/BMS270_2018/master/InstallTutorial.R")
```

In OSX and Ubuntu ensure that the folders RawData and RDS (downloaded by the above script) are in the same directory as <i>YourName.MicrobiomeTutorial.R</i>. <b>If you are running windows, the sequencing data was not automatically downloaded. Please manually download from this [link](https://github.com/jbisanz/BMS270_BMI219/raw/master/tutorialdata.zip), and decompress in the same directory as <i>YourName.MicrobiomeTutorial.R</i></b>

***

# Load packages

To ensure that you have successfully installed all packages, run the following commands one by one paying attention to the output. Warning messages about object masking are to be expected in this case but OK.

```{r, message=F, warning=F}
require(vegan) #Diversity functions for microbiome data
require(dada2) #Our denoising and primary sequence processing tool
require(phyloseq) #plotting and analysis tools for microbiome data
require(MicrobeR) #plotting and functions for microbiome data
require(DESeq2) #differential expression/feature abundance tool
require(genefilter) #useful to help filter tables

require(tidyverse) #a whole gang of packages and functions for data wrangling and graphing

require(DECIPHER) #for nucleotide alignments
require(phangorn) #phylogenetic tree tools
require(ggtree) #tree plotting tools

```

If any of these fail to load, try re-sourcing the InstallTutorial.R script above. It will only download the data once. If you still are unable to install any of the packages, please [email me](mailto:jordan.bisanz@ucsf.edu).

It is often quite handy to store intermediate files if R crashes, or for tracking of analyses. We will create a directory for this called RDS. We will store intermediate files in RDS format which allows them to be written and reloaded into R quickly and with minimal hard drive usage.

```{r}
dir.create("RDS")
```

Finally, it is a good idea to be aware of your environment and what versions of packages you have installed. You can do this with the following command. Your environment will not look exactly like mine, but this is a good thing to keep track of for when it comes to try to repeat analyses or publish your data.

```{r}
print(sessionInfo())
```

***

## You are now ready for the tutorial! Wait for class before moving on!

***

